File: storage/internal/chunk/open_chunk.go
// Package chunk — open-chunk state.
//
// Architecture note (A4-Metadata-As-Chunks, docs/02, docs/14 S11/S13/S14):
//
// NovaNas's "everything is chunks" invariant means the metadata service
// itself must live on chunks. Yet BadgerDB's value-log / WAL performs many
// small sequential appends and cannot wait for a full 4 MiB content-addressed
// chunk to fill before acknowledging fsync. To reconcile these, we extend
// the chunk engine with a second state:
//
//   OpenChunk:   mutable, append-only, UUID-identified, capacity-bounded.
//                Every append is fan-out-replicated to the chunk's replica
//                set (same quorum rules as sealed chunks). Reads of an open
//                chunk go through the owner replica.
//
//   SealedChunk: the classic content-addressed, immutable 4 MiB chunk. After
//                sealing, a ChunkID (SHA-256 of the final contents) is
//                computed and registered; subsequent writes are rejected.
//
// Transitions: Open -> (Append, Append, ..., Full || Timeout) -> Sealed.
//
// An open chunk may be sealed early (short sealed chunks are allowed for
// the WAL path). After sealing, the open-chunk UUID is retired and the
// chunk is addressable only by its content hash.
//
// This file defines the Go-side types and in-memory reference
// implementation. The real I/O path lives in the Rust data-plane
// (storage/dataplane/src/chunk). The gRPC surface is in
// proto/novanas/chunk/v1 (OpenChunk, AppendChunk, SealChunk).
package chunk

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"hash/crc32"
	"sync"
	"time"

	"github.com/azrtydxb/novanas/storage/internal/crypto"
)

// DefaultOpenChunkCapacity is the default capacity for an open chunk.
// 64 KiB is small enough to keep WAL latency low while amortising the
// seal-and-content-address cost over many entries. Tunable per pool.
const DefaultOpenChunkCapacity = 64 * 1024

// DefaultOpenChunkTimeout is the default wall-clock deadline after which an
// otherwise-idle open chunk is force-sealed even if not full.
const DefaultOpenChunkTimeout = 5 * time.Second

// OpenChunkID identifies an open (mutable) chunk via a random UUID. This is
// distinct from ChunkID (which is the content-addressed SHA-256 of a sealed
// chunk) because an open chunk's contents change with every append.
type OpenChunkID string

// NewOpenChunkID allocates a fresh random UUID-like identifier for an
// open chunk. Uses 16 random bytes rendered as hex (32 chars); sufficient
// to avoid collisions across a cluster's lifetime.
func NewOpenChunkID() (OpenChunkID, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("generating open-chunk id: %w", err)
	}
	return OpenChunkID(hex.EncodeToString(b[:])), nil
}

// OpenChunkState is the lifecycle state of an open chunk.
type OpenChunkState int

const (
	// OpenChunkOpen is the initial state: appends are accepted.
	OpenChunkOpen OpenChunkState = iota
	// OpenChunkSealed means the chunk has been sealed; it is now immutable
	// and addressable by its content-hash ChunkID.
	OpenChunkSealed
)

// OpenChunk is a mutable, append-only chunk in flight.
//
// Concurrency: OpenChunk is safe for concurrent Append/Seal callers. In
// the real data-plane the replication fan-out happens before Append
// returns; here we model just the local buffer state.
type OpenChunk struct {
	mu sync.Mutex

	id         OpenChunkID
	poolID     string
	capacity   int
	createdAt  time.Time
	lastAppend time.Time
	buf        []byte
	state      OpenChunkState
	sealedAs   ChunkID // set once state == OpenChunkSealed
}

// NewOpenChunk allocates an open chunk with the given capacity (bytes).
// capacity must be > 0 and <= ChunkSize.
func NewOpenChunk(poolID string, capacity int) (*OpenChunk, error) {
	if capacity <= 0 {
		return nil, fmt.Errorf("open-chunk capacity must be > 0, got %d", capacity)
	}
	if capacity > ChunkSize {
		return nil, fmt.Errorf("open-chunk capacity %d exceeds ChunkSize %d", capacity, ChunkSize)
	}
	id, err := NewOpenChunkID()
	if err != nil {
		return nil, err
	}
	now := time.Now()
	return &OpenChunk{
		id:         id,
		poolID:     poolID,
		capacity:   capacity,
		createdAt:  now,
		lastAppend: now,
		buf:        make([]byte, 0, capacity),
		state:      OpenChunkOpen,
	}, nil
}

// ID returns the open-chunk UUID.
func (o *OpenChunk) ID() OpenChunkID { return o.id }

// PoolID returns the pool this open chunk belongs to.
func (o *OpenChunk) PoolID() string { return o.poolID }

// Capacity returns the byte capacity of this open chunk.
func (o *OpenChunk) Capacity() int { return o.capacity }

// Len returns the current byte offset (how much has been appended).
func (o *OpenChunk) Len() int {
	o.mu.Lock()
	defer o.mu.Unlock()
	return len(o.buf)
}

// State returns the current lifecycle state.
func (o *OpenChunk) State() OpenChunkState {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.state
}

// SealedID returns the content-addressed ChunkID this open chunk was
// sealed as, or the empty string if it is still open.
func (o *OpenChunk) SealedID() ChunkID {
	o.mu.Lock()
	defer o.mu.Unlock()
	return o.sealedAs
}

// Errors.
var (
	ErrOpenChunkSealed       = errors.New("open chunk is already sealed")
	ErrOpenChunkFull         = errors.New("open chunk is full")
	ErrOpenChunkOffsetMismatch = errors.New("append offset does not match current length")
	ErrOpenChunkNotFound     = errors.New("open chunk not found")
)

// Append writes data at the given byte offset. The offset MUST equal the
// current length (appends are strictly append-only — no holes, no
// overwrites). This constraint lets replicas detect lost writes without
// a per-byte log.
//
// Returns ErrOpenChunkSealed if the chunk has been sealed,
// ErrOpenChunkFull if the append would exceed capacity,
// ErrOpenChunkOffsetMismatch if offset != current length.
func (o *OpenChunk) Append(offset int, data []byte) error {
	o.mu.Lock()
	defer o.mu.Unlock()

	if o.state == OpenChunkSealed {
		return ErrOpenChunkSealed
	}
	if offset != len(o.buf) {
		return fmt.Errorf("%w: got offset=%d, current length=%d",
			ErrOpenChunkOffsetMismatch, offset, len(o.buf))
	}
	if len(o.buf)+len(data) > o.capacity {
		return ErrOpenChunkFull
	}
	o.buf = append(o.buf, data...)
	o.lastAppend = time.Now()
	return nil
}

// Snapshot returns a copy of the current buffer contents. Intended for
// reads while the chunk is still open (e.g. crash recovery).
func (o *OpenChunk) Snapshot() []byte {
	o.mu.Lock()
	defer o.mu.Unlock()
	out := make([]byte, len(o.buf))
	copy(out, o.buf)
	return out
}

// Seal transitions Open -> Sealed. Computes the content hash over the
// currently-appended bytes, records it, and returns the resulting
// sealed Chunk. After Seal, further Append calls return
// ErrOpenChunkSealed.
//
// Sealing a zero-length open chunk is allowed (produces the hash of an
// empty byte slice); callers that want to reject empty seals should
// check Len() first.
func (o *OpenChunk) Seal() (*Chunk, error) {
	o.mu.Lock()
	defer o.mu.Unlock()

	if o.state == OpenChunkSealed {
		return nil, ErrOpenChunkSealed
	}
	sealedData := make([]byte, len(o.buf))
	copy(sealedData, o.buf)

	id := NewChunkID(sealedData)
	table := crc32.MakeTable(crc32.Castagnoli)
	c := &Chunk{
		ID:       id,
		Data:     sealedData,
		Checksum: crc32.Checksum(sealedData, table),
	}
	o.state = OpenChunkSealed
	o.sealedAs = id
	return c, nil
}

// SealEncrypted transitions Open -> Sealed with convergent encryption.
// The accumulated plaintext is encrypted under dk (32 bytes) before
// the content hash is computed, so the resulting ChunkID is
// SHA-256(ciphertext||auth_tag). This preserves dedup within the DK
// scope while guaranteeing data-at-rest encryption from the moment
// the chunk is sealed.
//
// For the open-chunk path we encrypt at seal time (rather than
// per-append) for simplicity: the buffer is small (64 KiB default)
// and seal-time encryption avoids the complexity of a streaming AEAD
// across many small appends.
func (o *OpenChunk) SealEncrypted(dk []byte) (*Chunk, error) {
	if len(dk) != crypto.ChunkKeySize {
		return nil, fmt.Errorf("open-chunk: dataset key must be %d bytes, got %d", crypto.ChunkKeySize, len(dk))
	}
	o.mu.Lock()
	defer o.mu.Unlock()

	if o.state == OpenChunkSealed {
		return nil, ErrOpenChunkSealed
	}
	sealedData := make([]byte, len(o.buf))
	copy(sealedData, o.buf)

	enc, err := crypto.EncryptChunk(dk, sealedData)
	if err != nil {
		return nil, fmt.Errorf("open-chunk: encrypt: %w", err)
	}
	id := ChunkID(hex.EncodeToString(enc.ChunkID[:]))
	table := crc32.MakeTable(crc32.Castagnoli)
	c := &Chunk{
		ID:            id,
		Data:          enc.Ciphertext,
		Checksum:      crc32.Checksum(enc.Ciphertext, table),
		Encrypted:     true,
		AuthTag:       enc.AuthTag,
		PlaintextHash: enc.PlaintextHash,
	}
	o.state = OpenChunkSealed
	o.sealedAs = id
	return c, nil
}

// ShouldSeal reports whether the open chunk should be sealed given its
// current fill level and an optional idle timeout. Sealing happens when
// the chunk is full (len == capacity) OR it has been idle longer than
// timeout (timeout == 0 disables the timeout check).
func (o *OpenChunk) ShouldSeal(timeout time.Duration) bool {
	o.mu.Lock()
	defer o.mu.Unlock()
	if o.state == OpenChunkSealed {
		return false
	}
	if len(o.buf) >= o.capacity {
		return true
	}
	if timeout > 0 && time.Since(o.lastAppend) >= timeout {
		return true
	}
	return false
}

// OpenChunkRegistry is an in-memory tracker for open chunks. The real
// data-plane keeps this in the Rust chunk engine; this Go-side registry
// is used by management-plane components and tests.
type OpenChunkRegistry struct {
	mu     sync.Mutex
	chunks map[OpenChunkID]*OpenChunk
}

// NewOpenChunkRegistry constructs an empty registry.
func NewOpenChunkRegistry() *OpenChunkRegistry {
	return &OpenChunkRegistry{chunks: make(map[OpenChunkID]*OpenChunk)}
}

// Open allocates a new open chunk and registers it.
func (r *OpenChunkRegistry) Open(poolID string, capacity int) (*OpenChunk, error) {
	c, err := NewOpenChunk(poolID, capacity)
	if err != nil {
		return nil, err
	}
	r.mu.Lock()
	r.chunks[c.id] = c
	r.mu.Unlock()
	return c, nil
}

// Get returns the open chunk with the given id, or ErrOpenChunkNotFound.
func (r *OpenChunkRegistry) Get(id OpenChunkID) (*OpenChunk, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	c, ok := r.chunks[id]
	if !ok {
		return nil, ErrOpenChunkNotFound
	}
	return c, nil
}

// Append is a convenience that looks up the open chunk and appends.
func (r *OpenChunkRegistry) Append(id OpenChunkID, offset int, data []byte) error {
	c, err := r.Get(id)
	if err != nil {
		return err
	}
	return c.Append(offset, data)
}

// Seal seals the open chunk, removes it from the open registry, and
// returns the resulting sealed Chunk.
func (r *OpenChunkRegistry) Seal(id OpenChunkID) (*Chunk, error) {
	c, err := r.Get(id)
	if err != nil {
		return nil, err
	}
	sealed, err := c.Seal()
	if err != nil {
		return nil, err
	}
	r.mu.Lock()
	delete(r.chunks, id)
	r.mu.Unlock()
	return sealed, nil
}

// SealEncrypted seals the open chunk with convergent encryption under
// the supplied dataset key, removes it from the registry, and returns
// the resulting encrypted sealed Chunk.
func (r *OpenChunkRegistry) SealEncrypted(id OpenChunkID, dk []byte) (*Chunk, error) {
	c, err := r.Get(id)
	if err != nil {
		return nil, err
	}
	sealed, err := c.SealEncrypted(dk)
	if err != nil {
		return nil, err
	}
	r.mu.Lock()
	delete(r.chunks, id)
	r.mu.Unlock()
	return sealed, nil
}

// List returns the ids of all currently-open chunks.
func (r *OpenChunkRegistry) List() []OpenChunkID {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]OpenChunkID, 0, len(r.chunks))
	for id := range r.chunks {
		out = append(out, id)
	}
	return out
}

File: storage/dataplane/src/chunk/open_chunk.rs
//! Open-chunk lifecycle (A4-Metadata-As-Chunks).
//!
//! Mirrors `storage/internal/chunk/open_chunk.go`. Open chunks are
//! mutable, append-only, UUID-identified buffers that seal into
//! content-addressed chunks. They provide WAL-style low-latency
//! semantics on top of the 4 MiB content-addressed chunk substrate so
//! that the metadata service (BadgerDB) can run on the chunk engine
//! without waiting for a full chunk to fill before fsync.
//!
//! Lifecycle: Open -> (Append, Append, ...) -> (Full | Timeout) -> Seal.
//!
//! This module is deliberately pure Rust (no SPDK dependencies) so it
//! builds in CI without the data-plane feature set. The real replication
//! fan-out plugs into ChunkEngine and lives in engine.rs; this module
//! only provides the per-chunk state machine and an in-memory registry.

use ring::digest;
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};
use uuid::Uuid;

use crate::backend::chunk_store::CHUNK_SIZE;

/// Default capacity of an open chunk (64 KiB). Tunable per pool.
pub const DEFAULT_OPEN_CHUNK_CAPACITY: usize = 64 * 1024;

/// Default idle timeout before an open chunk is force-sealed (5 s).
pub const DEFAULT_OPEN_CHUNK_TIMEOUT: Duration = Duration::from_secs(5);

/// Identifier for an open (mutable) chunk.
///
/// This is distinct from the content-addressed `ChunkId` because an
/// open chunk's contents change with every append.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct OpenChunkId(pub String);

impl OpenChunkId {
    /// Allocate a fresh random open-chunk id.
    pub fn new() -> Self {
        Self(Uuid::new_v4().simple().to_string())
    }
}

impl Default for OpenChunkId {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OpenChunkState {
    Open,
    Sealed,
}

/// Errors for the open-chunk lifecycle.
#[derive(Debug, thiserror::Error)]
pub enum OpenChunkError {
    #[error("open chunk is already sealed")]
    Sealed,
    #[error("open chunk is full (capacity {capacity} bytes)")]
    Full { capacity: usize },
    #[error("append offset {got} does not match current length {current}")]
    OffsetMismatch { got: usize, current: usize },
    #[error("open chunk not found: {0:?}")]
    NotFound(OpenChunkId),
    #[error("invalid capacity {0}: must be > 0 and <= {max}", max = CHUNK_SIZE)]
    InvalidCapacity(usize),
    #[error("chunk encryption failed")]
    EncryptFailed,
}

/// A single mutable, append-only open chunk.
#[derive(Debug)]
pub struct OpenChunk {
    id: OpenChunkId,
    pool_id: String,
    capacity: usize,
    buf: Vec<u8>,
    state: OpenChunkState,
    sealed_as: Option<[u8; 32]>,
    last_append: Instant,
}

impl OpenChunk {
    pub fn new(pool_id: impl Into<String>, capacity: usize) -> Result<Self, OpenChunkError> {
        if capacity == 0 || capacity > CHUNK_SIZE as usize {
            return Err(OpenChunkError::InvalidCapacity(capacity));
        }
        Ok(Self {
            id: OpenChunkId::new(),
            pool_id: pool_id.into(),
            capacity,
            buf: Vec::with_capacity(capacity),
            state: OpenChunkState::Open,
            sealed_as: None,
            last_append: Instant::now(),
        })
    }

    pub fn id(&self) -> &OpenChunkId {
        &self.id
    }

    pub fn pool_id(&self) -> &str {
        &self.pool_id
    }

    pub fn capacity(&self) -> usize {
        self.capacity
    }

    pub fn len(&self) -> usize {
        self.buf.len()
    }

    pub fn is_empty(&self) -> bool {
        self.buf.is_empty()
    }

    pub fn state(&self) -> OpenChunkState {
        self.state
    }

    pub fn sealed_as(&self) -> Option<&[u8; 32]> {
        self.sealed_as.as_ref()
    }

    /// Append data at the given offset; offset MUST equal current length.
    pub fn append(&mut self, offset: usize, data: &[u8]) -> Result<(), OpenChunkError> {
        if self.state == OpenChunkState::Sealed {
            return Err(OpenChunkError::Sealed);
        }
        if offset != self.buf.len() {
            return Err(OpenChunkError::OffsetMismatch {
                got: offset,
                current: self.buf.len(),
            });
        }
        if self.buf.len() + data.len() > self.capacity {
            return Err(OpenChunkError::Full {
                capacity: self.capacity,
            });
        }
        self.buf.extend_from_slice(data);
        self.last_append = Instant::now();
        Ok(())
    }

    /// Seal the chunk: compute SHA-256, freeze, return (sealed_id, data).
    pub fn seal(&mut self) -> Result<([u8; 32], Vec<u8>), OpenChunkError> {
        if self.state == OpenChunkState::Sealed {
            return Err(OpenChunkError::Sealed);
        }
        let digest_val = digest::digest(&digest::SHA256, &self.buf);
        let mut id = [0u8; 32];
        id.copy_from_slice(digest_val.as_ref());
        self.state = OpenChunkState::Sealed;
        self.sealed_as = Some(id);
        let out = std::mem::take(&mut self.buf);
        Ok((id, out))
    }

    /// Seal the chunk with convergent encryption under `dk` (32 bytes).
    /// Returns (chunk_id, ciphertext, auth_tag, plaintext_hash) where
    /// chunk_id = SHA-256(ciphertext || auth_tag). See
    /// `crate::crypto` for the scheme.
    pub fn seal_encrypted(
        &mut self,
        dk: &[u8],
    ) -> Result<([u8; 32], Vec<u8>, [u8; 16], [u8; 32]), OpenChunkError> {
        if self.state == OpenChunkState::Sealed {
            return Err(OpenChunkError::Sealed);
        }
        let enc = crate::crypto::encrypt_chunk(dk, &self.buf)
            .map_err(|_| OpenChunkError::EncryptFailed)?;
        self.state = OpenChunkState::Sealed;
        self.sealed_as = Some(enc.chunk_id);
        // Drop plaintext buffer.
        self.buf = Vec::new();
        Ok((
            enc.chunk_id,
            enc.ciphertext,
            enc.auth_tag,
            enc.plaintext_hash,
        ))
    }

    /// Whether this open chunk should be sealed now (full or idle past
    /// timeout). A zero timeout disables the idle check.
    pub fn should_seal(&self, timeout: Duration) -> bool {
        if self.state == OpenChunkState::Sealed {
            return false;
        }
        if self.buf.len() >= self.capacity {
            return true;
        }
        if !timeout.is_zero() && self.last_append.elapsed() >= timeout {
            return true;
        }
        false
    }
}

/// Replicator is the seam where CRUSH-driven fan-out is wired (#14).
///
/// On every `append`, the registry calls `replicate_append`; on
/// `seal`, `replicate_seal`. Implementations are responsible for:
///
///  - Picking replicas via CRUSH for the (pool_id, chunk_id) tuple.
///  - Fanning the operation to each replica synchronously — the
///    registry blocks the caller until acks return.
///  - Surfacing replica out-of-sync via the returned `Result`.
///
/// The registry uses a `NoopReplicator` when no replicator is set,
/// preserving the legacy single-node behavior. Production code wires
/// `GrpcReplicator` (lives in transport/, separate change).
pub trait Replicator: Send + Sync + std::fmt::Debug {
    /// Fan an append to replicas. Called *after* the local append
    /// succeeds; replication is best-effort if it fails the registry
    /// will surface the error and the caller can choose to roll back.
    fn replicate_append(
        &self,
        chunk: &OpenChunkId,
        offset: usize,
        data: &[u8],
    ) -> Result<(), OpenChunkError>;

    /// Fan a seal to replicas. ALL replicas must ack before the seal
    /// is considered complete (#14 acceptance: "all replicas must ack
    /// the seal before returning"). Returns Err on partial acks; the
    /// caller treats the seal as failed and the chunk stays open for
    /// retry.
    fn replicate_seal(&self, chunk: &OpenChunkId, sealed: &[u8]) -> Result<(), OpenChunkError>;
}

/// No-op Replicator: used in tests and single-node deployments.
#[derive(Debug, Default)]
pub struct NoopReplicator;

impl Replicator for NoopReplicator {
    fn replicate_append(
        &self,
        _chunk: &OpenChunkId,
        _offset: usize,
        _data: &[u8],
    ) -> Result<(), OpenChunkError> {
        Ok(())
    }

    fn replicate_seal(&self, _chunk: &OpenChunkId, _sealed: &[u8]) -> Result<(), OpenChunkError> {
        Ok(())
    }
}

/// In-memory registry of open chunks. Thread-safe via a single Mutex;
/// suitable for the modest number of concurrently-open chunks
/// (typically O(number of pools) for WAL traffic).
pub struct OpenChunkRegistry {
    inner: Mutex<HashMap<OpenChunkId, OpenChunk>>,
    replicator: Box<dyn Replicator>,
}

impl Default for OpenChunkRegistry {
    fn default() -> Self {
        Self {
            inner: Mutex::new(HashMap::new()),
            replicator: Box::new(NoopReplicator),
        }
    }
}

impl OpenChunkRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Build a registry with a custom Replicator (e.g. the gRPC
    /// fan-out used in multi-node deployments). The replicator is
    /// invoked synchronously inside `append` and `seal` so callers
    /// observe the same ordering as a single-node store.
    pub fn with_replicator(replicator: Box<dyn Replicator>) -> Self {
        Self {
            inner: Mutex::new(HashMap::new()),
            replicator,
        }
    }

    /// Allocate a new open chunk and return its id.
    pub fn open(
        &self,
        pool_id: impl Into<String>,
        capacity: usize,
    ) -> Result<OpenChunkId, OpenChunkError> {
        let c = OpenChunk::new(pool_id, capacity)?;
        let id = c.id().clone();
        self.inner.lock().unwrap().insert(id.clone(), c);
        Ok(id)
    }

    /// Append to the chunk with the given id, then fan the append
    /// out to replicas (#14). The fan-out is synchronous so the
    /// caller observes a single linearisable order — when this
    /// returns Ok the data is durable on every selected replica.
    pub fn append(
        &self,
        id: &OpenChunkId,
        offset: usize,
        data: &[u8],
    ) -> Result<usize, OpenChunkError> {
        let chunk_len = {
            let mut g = self.inner.lock().unwrap();
            let c = g
                .get_mut(id)
                .ok_or_else(|| OpenChunkError::NotFound(id.clone()))?;
            c.append(offset, data)?;
            c.len()
        };
        // Replicate outside the lock — replicator may issue gRPC.
        self.replicator.replicate_append(id, offset, data)?;
        Ok(chunk_len)
    }

    /// Seal the chunk locally, fan the seal out to replicas, and
    /// remove it from the registry once all replicas ack. If any
    /// replica fails to ack, the chunk is reinstated so the caller
    /// can retry (#14 acceptance).
    pub fn seal(&self, id: &OpenChunkId) -> Result<([u8; 32], Vec<u8>), OpenChunkError> {
        let mut c = {
            let mut g = self.inner.lock().unwrap();
            g.remove(id)
                .ok_or_else(|| OpenChunkError::NotFound(id.clone()))?
        };
        let result = c.seal()?;
        if let Err(err) = self.replicator.replicate_seal(id, &result.1) {
            // Roll back: the local seal is irreversible (state machine
            // only goes one way), but we put the chunk back so a retry
            // can re-issue the seal RPC against the same content.
            self.inner.lock().unwrap().insert(id.clone(), c);
            return Err(err);
        }
        Ok(result)
    }

    pub fn len(&self) -> usize {
        self.inner.lock().unwrap().len()
    }

    pub fn is_empty(&self) -> bool {
        self.inner.lock().unwrap().is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn append_and_seal_roundtrip() {
        let mut c = OpenChunk::new("pool-meta", 32).unwrap();
        c.append(0, b"hello ").unwrap();
        c.append(6, b"world").unwrap();
        assert_eq!(c.len(), 11);
        let (id, data) = c.seal().unwrap();
        assert_eq!(data, b"hello world".to_vec());
        // SHA-256 of "hello world".
        let expected = digest::digest(&digest::SHA256, b"hello world");
        assert_eq!(&id, <&[u8; 32]>::try_from(expected.as_ref()).unwrap());
        assert_eq!(c.state(), OpenChunkState::Sealed);
    }

    #[test]
    fn offset_mismatch_rejected() {
        let mut c = OpenChunk::new("p", 32).unwrap();
        c.append(0, b"abc").unwrap();
        match c.append(0, b"x") {
            Err(OpenChunkError::OffsetMismatch { got: 0, current: 3 }) => {}
            other => panic!("expected offset mismatch, got {:?}", other),
        }
    }

    #[test]
    fn full_rejected() {
        let mut c = OpenChunk::new("p", 4).unwrap();
        c.append(0, b"abcd").unwrap();
        assert!(matches!(
            c.append(4, b"e"),
            Err(OpenChunkError::Full { .. })
        ));
    }

    #[test]
    fn sealed_is_immutable() {
        let mut c = OpenChunk::new("p", 32).unwrap();
        c.append(0, b"data").unwrap();
        let _ = c.seal().unwrap();
        assert!(matches!(c.append(4, b"x"), Err(OpenChunkError::Sealed)));
        assert!(matches!(c.seal(), Err(OpenChunkError::Sealed)));
    }

    #[test]
    fn invalid_capacity() {
        assert!(matches!(
            OpenChunk::new("p", 0),
            Err(OpenChunkError::InvalidCapacity(0))
        ));
        assert!(matches!(
            OpenChunk::new("p", (CHUNK_SIZE as usize) + 1),
            Err(OpenChunkError::InvalidCapacity(_))
        ));
    }

    #[test]
    fn should_seal_full() {
        let mut c = OpenChunk::new("p", 4).unwrap();
        assert!(!c.should_seal(Duration::from_secs(60)));
        c.append(0, b"abcd").unwrap();
        assert!(c.should_seal(Duration::ZERO));
    }

    #[test]
    fn seal_encrypted_roundtrip() {
        let dk = [0x42u8; 32];
        let mut c = OpenChunk::new("p", 64).unwrap();
        c.append(0, b"hello").unwrap();
        c.append(5, b" world").unwrap();
        let (id, ct, tag, ph) = c.seal_encrypted(&dk).unwrap();
        assert_eq!(c.state(), OpenChunkState::Sealed);
        assert_ne!(
            ct,
            b"hello world".to_vec(),
            "ciphertext must differ from plaintext"
        );
        // id = SHA-256(ct || tag)
        let mut ctx = digest::Context::new(&digest::SHA256);
        ctx.update(&ct);
        ctx.update(&tag);
        let expected = ctx.finish();
        assert_eq!(&id[..], expected.as_ref());
        // Roundtrip.
        let got = crate::crypto::decrypt_chunk(&dk, &ct, &tag, &ph).unwrap();
        assert_eq!(got, b"hello world");
    }

    #[test]
    fn registry_lifecycle() {
        let reg = OpenChunkRegistry::new();
        let id = reg.open("pool", 64).unwrap();
        let after = reg.append(&id, 0, b"abc").unwrap();
        assert_eq!(after, 3);
        let after2 = reg.append(&id, 3, b"def").unwrap();
        assert_eq!(after2, 6);
        let (_, data) = reg.seal(&id).unwrap();
        assert_eq!(data, b"abcdef".to_vec());
        // After seal it must be gone.
        assert!(matches!(
            reg.append(&id, 0, b"x"),
            Err(OpenChunkError::NotFound(_))
        ));
    }
}
